internal/controllers/machineset/machineset_controller.go

```diff
@@ -966,6 +994,146 @@ func (r *Reconciler) getMachineNode(ctx context.Context, cluster *clusterv1.Clus
+			return result, errors.Wrap(err, "failed to perform pre-flight checks")
+		}
+	}
+
+	// Run the kubeadm version skew preflight check.
+	if !skippable.Has(clusterv1.MachineSetPreflightCheckKubeadm) {
+		if result := r.kubeadmVersionPreflightCheck(ctx, *cpVersion, ms); !result.IsZero() {
+			return result, nil
+		}
+	}
+
+	// All pre-flight checks passed.
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) controlPlaneStablePreflightCheck(ctx context.Context, controlPlane *unstructured.Unstructured) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	cpKlogRef := klog.KRef(controlPlane.GetNamespace(), controlPlane.GetName())
+
+	// Check that the control plane is not provisioning.
+	isProvisioning, err := contract.ControlPlane().IsProvisioning(controlPlane)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform control plane pre-flight check: failed to check if ControlPlane %s is provisioning", cpKlogRef)
+	}
+	if isProvisioning {
+		log.Info(fmt.Sprintf("Pre-flight check failed: ControlPlane %s is provisioning", cpKlogRef))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	// Check that the control plane is not upgrading.
+	isUpgrading, err := contract.ControlPlane().IsUpgrading(controlPlane)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform control plane pre-flight check: failed to check if the ControlPlane %s is upgrading", cpKlogRef)
+	}
+	if isUpgrading {
+		log.Info(fmt.Sprintf("Pre-flight check failed: ControlPlane %s is upgrading", cpKlogRef))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) kubernetesVersionPreflightCheck(ctx context.Context, cpVersion string, ms *clusterv1.MachineSet) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	// If the version is not defined on the MachineSet then we cannot check for the version skew policies.
+	// Return early.
+	if ms.Spec.Template.Spec.Version == nil {
+		return ctrl.Result{}, nil
+	}
+	msVersion := *ms.Spec.Template.Spec.Version
+
+	cpSemver, err := semver.ParseTolerant(cpVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform kubernetes version skew pre-flight check: failed to parse ControlPlane version %s", cpVersion)
+	}
+	msSemver, err := semver.ParseTolerant(msVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform kubernetes version skew pre-flight check: failed to parse MachineSet version %s", msVersion)
+	}
+
+	// Check the Kubernetes version skew policy.
+	// => MS minor version cannot be greater than the Control Plane minor version.
+	// => MS minor version cannot be older than 2 minor version of Control Plane.
+	// Kubernetes skew policy: https://kubernetes.io/releases/version-skew-policy/#kubelet
+	floorVersion := semver.Version{
+		Major: cpSemver.Major,
+		Minor: cpSemver.Minor - 2,
```

Comment:
Should this be -3? I think the recent KEP was merged

Might be good to start with -3 so that 1.5.0+ already support th enew policy

Reply:
That might depend on the Control Plane version right? If control plane version is >= X then it will support n-3 worker nodes if not it will only support n-2 worker nodes?

Reply:
Yeah probably. Just not sure if I want to care about the details and just use n-3 for validation :D

Reply:
I think it's fine to use - 2 for now. But this is something we should check when we start bumping to 1.28. Could you already create the Task for 1.28 and add a reminder as a release-specific sub-task?

Reply:
Created the v1.28 issue and added a release specific task under it: https://github.com/kubernetes-sigs/cluster-api/issues/8708